src/lib/Libdb/db_postgres_attr.c

```diff
@@ -91,123 +167,147 @@ struct pg_array {
  *
  */
 int
-convert_array_to_db_attr_list(char *raw_array, pbs_db_attr_list_t *attr_list)
+dbarray_2_attrlist(char *raw_array, pbs_db_attr_list_t *attr_list)
 {
 	int i;
 	int j;
 	int rows;
-	char *p;
 	char *attr_name;
 	char *attr_value;
-	pbs_db_attr_info_t *attrs = NULL;
+	char *attr_flags;
+	char *attr_resc;
+	svrattrl *pal;
 	struct pg_array *array = (struct pg_array *) raw_array;
 	struct str_data *val = (struct str_data *)(raw_array + sizeof(struct pg_array));
 
-	if (ntohl(array->ndim) == 0) {
-		attr_list->attributes = attrs;
-		attr_list->attr_count = 0;
+	CLEAR_HEAD(attr_list->attrs);
+	attr_list->attr_count = 0;
+
+	if (ntohl(array->ndim) == 0)
 		return 0;
-	}
 
-	if (ntohl(array->ndim) > 1 || ntohl(array->elemtype) != TEXTOID) {
+	if (ntohl(array->ndim) > 1 || ntohl(array->elemtype) != TEXTOID)
 		return -1;
-	}
 
 	rows = ntohl(array->size);
-	attrs = malloc(sizeof(pbs_db_attr_info_t)*rows/2);
-	if (!attrs)
-		return -1;
-
-	attr_list->attributes = attrs;
 
 	for (i=0, j = 0; j < rows; i++, j+=2) {
+
+		attr_resc = NULL;
+		attr_value = NULL;
+		
 		attr_name = val->str;
 		val = (struct str_data *)((char *) val->str + ntohl(val->len));
 
-		attr_value = val->str;
+		attr_flags = val->str;
 		val = (struct str_data *)((char *) val->str + ntohl(val->len));
 
-		if (attr_name) {
-			attrs[i].attr_name[sizeof(attrs[i].attr_name) -1] = '\0';
-			strncpy(attrs[i].attr_name, attr_name, sizeof(attrs[i].attr_name));
-			if ((p = strchr(attrs[i].attr_name, '.'))) {
-				*p = '\0';
-				attrs[i].attr_resc[sizeof(attrs[i].attr_resc) -1] = '\0';
-				strncpy(attrs[i].attr_resc, p + 1, sizeof(attrs[i].attr_resc));
-			} else
-				attrs[i].attr_resc[0] = '\0';
-		} else
-			attrs[i].attr_name[0] = 0;
-
-		if (attr_value && (p = strchr(attr_value, '.'))) {
-			*p ='\0';
-			attrs[i].attr_flags = atol(attr_value);
-			attrs[i].attr_value = strdup(p + 1);
-		} else {
-			attrs[i].attr_flags = 0;
-			attrs[i].attr_value = NULL;
+		if ((attr_resc = strchr(attr_name, '.'))) {
+			*attr_resc = '\0';
+			attr_resc++;
 		}
+
+		if ((attr_value = strchr(attr_flags, '.'))) {
+			*attr_value = '\0';
+			attr_value++;
+		}
+
+		if (!(pal = make_attr(attr_name, attr_resc, attr_value, atol(attr_flags))))
```

Comment:
Why? Per the man pages: The atol subroutine is equivalent to the strtol subroutine where the value of the EndPointer parameter is a null pointer and the Base parameter is a value of 10.

Reply:
We usually tend to avoid using ato* functions. These functions return numerical value even in case of failures (like when the input itself is non numeric)

Reply:
ah ok - the data is not from user input, from the database actually, so is entirely numeric - it can get corrupted in very bad cases of course, but there is not much to salvage at that point anyway. Nevertheless, i changed to strtol() - please take a peek